File: apps/ose-be/tests/unit/Steps/NatsConfigSteps.fs
module OseBe.Tests.Unit.Steps.NatsConfigSteps

open TickSpec
open Xunit
open OseBe.Infrastructure.NatsClient

let mutable private readEnvironment: string -> string = fun _ -> null
let mutable private caughtException: exn option = None

[<Given>]
let ``OSE_BE_NATS_URL is unset`` () =
    readEnvironment <- fun _ -> null
    caughtException <- None

[<When>]
let ``ose-be reads its messaging configuration`` () =
    caughtException <-
        try
            requireNatsUrlWith readEnvironment |> ignore
            None
        with ex ->
            Some ex

[<Then>]
let ``startup aborts with a clear missing-variable error`` () =
    Assert.True(caughtException.IsSome)
    Assert.Contains("OSE_BE_NATS_URL", caughtException.Value.Message)

File: apps/ose-be-e2e/steps/nats-config.steps.ts
import { spawnSync, type SpawnSyncReturns } from "node:child_process";
import { expect } from "@playwright/test";
import { createBdd } from "playwright-bdd";

const { Given, When, Then } = createBdd();

let childEnvironment: NodeJS.ProcessEnv = {};
let startupResult: SpawnSyncReturns<string> | undefined;

Given("OSE_BE_NATS_URL is unset", async () => {
  childEnvironment = { ...process.env, OSE_BE_NATS_URL: "" };
  expect(childEnvironment["OSE_BE_NATS_URL"]).toBe("");
});

When("ose-be reads its messaging configuration", async () => {
  startupResult = spawnSync(
    "dotnet",
    ["run", "--project", "../ose-be/src/OseBe/OseBe.fsproj", "--no-build", "--configuration", "Release"],
    { cwd: process.cwd(), env: childEnvironment, encoding: "utf8", timeout: 30_000 },
  );
});

Then("startup aborts with a clear missing-variable error", async () => {
  expect(startupResult).toBeDefined();
  expect(startupResult!.status).not.toBe(0);
  expect(`${startupResult!.stdout}\n${startupResult!.stderr}`).toContain("OSE_BE_NATS_URL");
});
